lib/spack/spack/url_buildcache.py

```diff
@@ -1049,6 +1077,131 @@ def check_mirror_for_layout(mirror: spack.mirrors.mirror.Mirror):
+
+    if not aws:
+        tty.warn("Failed to use aws s3 sync to retrieve specs, falling back to parallel fetch")
+        return file_list, read_fn
+
+    def file_read_method(manifest_path: str) -> URLBuildcacheEntry:
+        cache_entry = cache_class(mirror_url=url, allow_unsigned=True)
+        cache_entry.read_manifest(manifest_url=f"file://{manifest_path}")
+        return cache_entry
+
+    include_pattern = cache_class.get_buildcache_component_include_pattern(component_type)
+
+    sync_command_args = [
+        "s3",
+        "sync",
+        "--exclude",
+        "*",
+        "--include",
+        include_pattern,
+        url,
+        tmpspecsdir,
+    ]
+
+    tty.debug(f"Using aws s3 sync to download manifests from {url} to {tmpspecsdir}")
+
+    try:
+        aws(*sync_command_args, output=os.devnull, error=os.devnull)
+        file_list = fsys.find(tmpspecsdir, [include_pattern])
+        read_fn = file_read_method
+    except Exception:
+        tty.warn("Failed to use aws s3 sync to retrieve specs, falling back to parallel fetch")
+
+    return file_list, read_fn
+
+
+def _entries_from_cache_fallback(
+    url: str, tmpspecsdir: str, component_type: Optional[BuildcacheComponent] = None
+):
+    """Use spack.util.web module to get a list of all the specs at the remote url.
+
+    Args:
+        url: Base url of mirror (location of spec files)
+        tmpspecsdir: path to temporary directory to use for writing files
+        component_type: type of buildcache component to sync (spec, index, key, etc.)
+
+    Return:
+        The list of complete spec file urls and a function that can read each one from its
+            remote location (also using the spack.util.web module).
+    """
+    read_fn = None
+    file_list = None
+
+    cache_class = get_url_buildcache_class(layout_version=CURRENT_BUILD_CACHE_LAYOUT_VERSION)
+
+    def url_read_method(manifest_url: str) -> URLBuildcacheEntry:
+        cache_entry = cache_class(mirror_url=url, allow_unsigned=True)
+        cache_entry.read_manifest(manifest_url)
+        return cache_entry
+
+    try:
+        file_list = [
+            url_util.join(url, entry)
+            for entry in web_util.list_url(url, recursive=True)
+            if fnmatch.fnmatch(
+                entry, cache_class.get_buildcache_component_include_pattern(component_type)
+            )
+        ]
+        read_fn = url_read_method
+    except Exception as err:
+        # If we got some kind of S3 (access denied or other connection error), the first non
+        # boto-specific class in the exception is Exception.  Just print a warning and return
+        tty.warn(f"Encountered problem listing packages at {url}: {err}")
+
+    return file_list, read_fn
+
+
+def get_entries_from_cache(
+    url: str, tmpspecsdir: str, component_type: Optional[BuildcacheComponent] = None
+):
+    """Get a list of all the spec files in the mirror and a function to
```

Comment:
I think this whole docstring probably needs a pass over it, as this method isn't specific to specs anymore, but deals in manifests now.